### PR TITLE
Add LDAP server + CA to staging

### PR DESCRIPTION
**Pull request by Logiraptor**
This PR adds a new VM to the ES vpc, configures it to run OpenLDAP, and registers users with LDAP. 
Each user has a client cert generated which can be used to log in through portal.
This PR also configures portal to connect to the new LDAP server with the correct certs and credentials.

### CODE REVIEW

**Comment by Logiraptor** [comment]
Still testing this, don't merge it yet

**Comment by Logiraptor** [comment]
At a high level: This adds layer2-ldap-server, updates layer4-om-config to read from that output, and makes a few extra changes to support re-use (like adding a variable for port in the elb/create module)

File: layers/layer2-ldap-server/main.tf [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  port              = 636
+}
+
+resource "aws_elb_attachment" "ldap_attach" {
+  elb      = "${module.ldap_elb.my_elb_id}"
+  instance = "${module.ldap_host.instance_id}"
+}
+
+output "password" {
+  value = "${module.ldap_configure.password}"
```
Comment (fbehrens51) [inline layers/layer2-ldap-server/main.tf]: sensitive=true

File: layers/layer2-ldap-server/modules/ldap-server/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+output "basedn" {
+  value = "${local.basedn}"
+}
+
+output "role_attr" {
+  value = "role"
+}
+
+output "password" {
+  value = "${random_string.ldap_password.result}"
```
Comment (fbehrens51) [inline layers/layer2-ldap-server/modules/ldap-server/outputs.tf]: sensitive=true

File: layers/layer0-paperwork/modules/server_cert/main.tf [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+resource "tls_locally_signed_cert" "client_cert" {
+  cert_request_pem   = "${tls_cert_request.client_cert_request.cert_request_pem}"
+  ca_key_algorithm   = "RSA"
+  ca_private_key_pem = "${var.ca_private_key_pem}"
+  ca_cert_pem        = "${var.ca_cert_pem}"
+
+  allowed_uses = [
+    "digital_signature",
+    "server_auth",
+    "client_auth",
```
Comment (fbehrens51) [inline layers/layer0-paperwork/modules/server_cert/main.tf]: similar to above, does the server cert need client_auth?  I'd also rename the resources from client_... to server_... to be clearer

File: layers/layer0-paperwork/modules/ca_cert/main.tf [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    common_name  = "${var.env_name} Root CA"
+    organization = "${var.env_name}"
+  }
+
+  validity_period_hours = 8760 # 365 days
+
+  early_renewal_hours = 4320 # 180 days
+
+  allowed_uses = [
+    "digital_signature",
```
Comment (fbehrens51) [inline layers/layer0-paperwork/modules/ca_cert/main.tf]: does the root CA need client_auth and server_auth?

File: layers/layer0-paperwork/modules/paperwork/outputs.tf [hunk 5]
```diff
@@ -13,3 +13,36 @@ output "es_vpc_id" {
...
+output "root_ca_cert" {
+  value = "${module.ca_cert.cert_pem}"
+}
+
+output "ldap_server_cert" {
+  value = "${module.ldap_server_cert.cert_pem}"
+}
+
+output "ldap_server_key" {
+  value = "${module.ldap_server_cert.private_key_pem}"
```
[the remark this answers is not in the document]
Comment (Logiraptor) [reply layers/layer0-paperwork/modules/paperwork/outputs.tf]: done

File: layers/layer0-paperwork/modules/paperwork/outputs.tf [hunk 6]
```diff
@@ -13,3 +13,36 @@ output "es_vpc_id" {
...
+}
+
+output "ldap_client_cert" {
+  value = "${module.ldap_client_cert.cert_pem}"
+}
+
+output "ldap_client_key" {
+  value = "${module.ldap_client_cert.private_key_pem}"
```
[the remark this answers is not in the document]
Comment (Logiraptor) [reply layers/layer0-paperwork/modules/paperwork/outputs.tf]: done

File: layers/layer0-paperwork/modules/paperwork/outputs.tf [hunk 7]
```diff
@@ -13,3 +13,36 @@ output "es_vpc_id" {
...
+}
+
+output "user_certs" {
+  value = "${zipmap(data.template_file.usernames.*.rendered, tls_locally_signed_cert.user_pki_cert.*.cert_pem)}"
+}
+
+output "user_private_keys" {
+  value = "${zipmap(data.template_file.usernames.*.rendered, tls_private_key.user_pki_cert_private_key.*.private_key_pem)}"
+}
```
[the remark this answers is not in the document]
Comment (Logiraptor) [reply layers/layer0-paperwork/modules/paperwork/outputs.tf]: done

File: layers/layer0-paperwork/modules/paperwork/user_cert.tf [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+resource "tls_locally_signed_cert" "user_pki_cert" {
+  count              = "${length(var.users)}"
+  cert_request_pem   = "${tls_cert_request.user_pki_cert_request.*.cert_request_pem[ count.index ]}"
+  ca_key_algorithm   = "RSA"
+  ca_private_key_pem = "${module.ca_cert.private_key_pem}"
+  ca_cert_pem        = "${module.ca_cert.cert_pem}"
+
+  allowed_uses = [
+    "digital_signature",
+    "server_auth",
```
[the remark this answers is not in the document]
Comment (Logiraptor) [reply layers/layer0-paperwork/modules/paperwork/user_cert.tf]: done

File: layers/layer2-ldap-server/modules/ldap-server/outputs.tf (hunk 2 above)
Comment (Logiraptor) [reply layers/layer2-ldap-server/modules/ldap-server/outputs.tf]: done

File: layers/layer2-ldap-server/main.tf (hunk 1 above)
Comment (Logiraptor) [reply layers/layer2-ldap-server/main.tf]: done

File: layers/layer0-paperwork/modules/ca_cert/main.tf (hunk 4 above)
Comment (Logiraptor) [reply layers/layer0-paperwork/modules/ca_cert/main.tf]: removed them

File: layers/layer0-paperwork/modules/server_cert/main.tf (hunk 3 above)
Comment (Logiraptor) [reply layers/layer0-paperwork/modules/server_cert/main.tf]: I renamed those resources, and split out a second client_cert module. One with server_auth and signature only, the other with client_auth and signature only.

**Pull request merged**